src/Core/OrganizationFeatures/AuthorizationHandlers/OrganizationOperations.cs

```diff
@@ -0,0 +1,10 @@
+﻿using Microsoft.AspNetCore.Authorization.Infrastructure;
+
+namespace Bit.Core.OrganizationFeatures.AuthorizationHandlers;
+
+public class OrganizationOperationRequirement : OperationAuthorizationRequirement { }
+
+public static class OrganizationOperations
+{
+    public static readonly OrganizationOperationRequirement ReadAllGroups = new() { Name = "ReadAllGroups" };
```

Comment:
My gut is saying that we are already creating a weird web of requirements.

I see why you put `ReadAllGroups` on the org, but under what conditions, exactly, can someone read a single group, but not all?

I would love to figure out a way to Scope a user can read, so that when we are asked to index them, we return only the groups they are approved for, rather than just balking that they asked to read all.

Maybe that's too much of a digression from current behavior, but I don't think so.

This isn't really idiomatic here, but the lack of scoping has bugged me for a long time, and having `...all...` requirements makes things quite confusing to me, at the least multiplying your requirement count by ~1.5x

Reply:
> under what conditions, exactly, can someone read a single group, but not all?

This was an interesting find from this PR - once I broke down the logic for each discrete operation, it turns out that more people have permission to read _all_ groups than to read a _particular_ group. When presumably it should be the reverse.

I don't intend to resolve that inconsistency in this PR, but I do think these changes will help us identify and fix this in the future.

> [...] so that when we are asked to index them, we return only the groups they are approved for, rather than just balking that they asked to read all.

This doesn't seem to be supported by `IAuthorizationService` and I can't find any firm guidance on best practices here - only a few suggestions to just filter the list of resources at the database level (which we do in some places, e.g. [for ciphers](https://github.com/bitwarden/server/blob/master/src/Sql/Vault/dbo/Functions/UserCipherDetails.sql)).

Doing it at the database level duplicates the authorization logic, but it does seem like the most performant answer given the potential number of records involved. I'm not sure how you'd implement this at the code level without sending back large amounts of data from the database and then doing all your heavy lifting on the server. Maybe you could with EF, but Dapper? 

This is a genuine "I don't know", so suggestions here are welcome.

> Maybe that's too much of a digression from current behavior, but I don't think so.

It could be: at the moment, when the client is asking for _all_ of a resource, that's intentional, and it expects its request to be granted in full or denied in full. Not partially fulfilled with no warning.

For example: you try to do a whole vault export and instead of giving you all ciphers, the server gives you only the ciphers you're allowed to access. You think you've successfully backed up your vault, but you haven't.

Maybe that points to a bug on the front-end (i.e. the UI/UX is misleading the user re: their level of permissions and what they can do) and that's not the responsibility of the server. But it might cause a mismatch of expectations between client & server if we made this change today.

Reply:
> Maybe the biggest issue here is that now in order to determine if I can access this entity, I need to know to grab that other entity and check if it has permission to access all of these children. [...] It's not OK to have to have our entity hierarchy currently loaded into your head to be able to permission appropriately.

I agree, but as a practical matter I don't know what a better interface would be.

For example:

```cs
_authorizationService.AuthorizeAsync(User, typeof(Group), GroupOperations.Read);
```

Would be nice and consistent if orgs didn't exist - basically using the resource's type to represent "all resources of this type".  But as a practical matter, we need to be able to look up the org in the handler to decide whether the user is authorized, so we need to get the orgId in there somehow.

Like:

```cs
_authorizationService.AuthorizeAsync(User, typeof(Group), GroupOperations.Read, organizationId);
```

...but is that any better? And now the generic `authorizeAsync` interface has been customized to the particular auth logic of 1 particular resource.

I thought about using a "special" Group object to represent "all groups for this org", e.g.

```cs
_authorizationService.AuthorizeAsync(User, new Group() { OrganizationId = orgId }, GroupOperations.ReadAll);
```

Although now you have to know how to construct the Group object, which is also breaking the abstraction. Wrapping that in a static factory method provides a bit more guidance for devs:

```cs
_authorizationService.AuthorizeAsync(User, Group.All(orgId), GroupOperations.ReadAll);
```

But ultimately still feels hacky.

We could go in a totally different direction for this PR and just use a separate service per resource, allowing a more bespoke interface for each operation so we can just ask directly for the information we need:

```cs
_groupAuthorizationService.CreateAsync(group);
_groupAuthorizationService.ReadAsync(group);
_groupAuthorizationService.UpdateAsync(group);
_groupAuthorizationService.DeleteAsync(group);
_groupAuthorizationService.ReadAllAsync(orgId);
```

which might be OK. We'd have to reconsider whether `IAuthorizationHandler` has an benefit for our use case which we'd be missing out on (I don't know).

That's my stream of consciousness - what are you thinking about here, in more concrete terms?